src/tamm/tensor_impl.hpp

```diff
@@ -1189,8 +1194,9 @@ class DenseTensorImpl: public TensorImpl<T> {
             upcxx::global_ptr<uint8_t> remote_addr =
               gptrs_[t.rank] + (t.offset * elem_sz) + (tile_offset * elem_sz);
             uint8_t* local_addr = ((uint8_t*) buf) + (next++ * elem_sz);
-            upcxx::rget(remote_addr, local_addr, elem_sz).wait();
+            upcxx::rget(remote_addr, local_addr, elem_sz, upcxx::operation_cx::as_promise(p));
           }
+    p.finalize().wait();
```

Comment:
Similarly here, please do not call fine-grained `upcxx::rget` in a loop like this.
just push the addresses into vectors, and then make an `rget_regular()` call.